src/cashweb/wallet/index.ts

```diff
@@ -296,64 +290,68 @@ export class Wallet {
-    console.log(`Checking UTXO ${utxoId}`)
-    const client = await this.electrumClientPromise
-    assert(client, 'missing client in fixOutpointId')
-    const scriptHash = toElectrumScriptHash(utxo.address)
-    const elOutputs = await client.request(
-      'blockchain.scripthash.listunspent',
-      scriptHash,
-    )
-    assert(
-      elOutputs,
-      `Null response from blockchain.scripthash.listunspent for ${scriptHash}`,
-    )
-    if (elOutputs instanceof Error) {
-      return false
-    }
-    assert(
-      elOutputs instanceof Array,
-      'output of blockchain.scripthash.listunspent was not an array!',
+  async checkAreOutpointsUnspent(utxos: Utxo[]): Promise<boolean[]> {
+    const utxoIds = utxos.map(calcUtxoId)
+    console.log(`Checking UTXOs ${utxoIds}`)
+    const chronikClient = this.chronikClient
+    assert(chronikClient, 'missing client in checkAreOutpointsUnspent')
+    const utxoStates = await chronikClient.validateUtxos(
+      utxos.map(utxo => ({
+        txid: utxo.txId,
+        outIdx: utxo.outputIndex,
+      })),
     )
-    const unspentOutputs = elOutputs as ElectrumListUnspentUtxo[]
-    if (
-      !unspentOutputs.some(output => {
-        return (
-          output.tx_hash === utxo.txId && output.tx_pos === utxo.outputIndex
-        )
-      })
-    ) {
-      console.log(`UTXO missing on-chain, deleting missing UTXO ${utxoId}`)
-      this.storage.deleteById(utxoId)
+    return utxoStates.map((utxoState, idx) => {
+      switch (utxoState.state) {
+        case 'UNSPENT':
+          console.log('UTXO is valid', utxoIds[idx])
+          return true
+        case 'SPENT':
+          console.log('UTXO spent on-chain, deleting spent UTXO', utxoIds[idx])
+          break
+        case 'NO_SUCH_TX':
+          console.log(
+            'Invalid UTXO: tx does not exist on-chain, deleting invalid UTXO',
+            utxoIds[idx],
+          )
+          break
+        case 'NO_SUCH_OUTPUT':
+          console.log(
+            "Invalid UTXO: tx exists, but output doesn't, deleting invalid UTXO",
+            utxoIds[idx],
+          )
+          break
+      }
       return false
-    }
-    console.log(`UTXO is valid ${utxoId}`)
-    return true
+    })
   }
 
-  async fixUtxo(utxo: Utxo) {
-    const utxoId = calcUtxoId(utxo)
+  async fixUtxos(utxos: Utxo[]) {
     // TODO: This needs some thought to ensure we do not delete outpoints that
     // are possibly just out of sync with the mempool.
+    const utxoIds = utxos.map(calcUtxoId)
     try {
       // Unfreeze UTXO if confirmed to be unspent else delete
       // WARNING: This is not thread-safe, do not call when others hold the UTXO
-      const utxo = await this.storage.getById(utxoId)
-      if (!utxo) {
-        console.log(`Missing UTXO for ${utxoId}`)
-        return
-      }
-      if (!(await this.checkOutpoint(utxo))) {
-        return
-      }
-      await this.unfreezeUtxo(utxoId)
+      const storageUtxos = utxoIds.flatMap(utxoId => {
```

Comment:
After loading all the messages, there might be some UTXOs that no longer exist on-chain. Those ones need to be deleted from local state at some point. This function no longer deletes those ones because it doesn't call checkUtxosAreUnspent, and also that function doesn't delete them.